### PR TITLE
Fix compiler error on gcc 4.5.2.

### PR DESCRIPTION
**Pull request by mschulz-at-hilscher**
## Description

When compiling with:
```
arm-none-eabi-gcc.exe (Sourcery G++ Lite 2011.03-42) 4.5.2
```
We get the following error:
```
constant_time_impl.h:47:13: error: expected [error|warning|ignored] after '#pragma GCC diagnostic'
```

using this commit, we activat the pragma only for GCC 5+

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [X] **changelog** not required
- [X] **backport** not required
- [X] **tests** not required

### CODE REVIEW

File: library/constant_time_impl.h [hunk 1]
```diff
@@ -31,7 +31,7 @@
  * Disable -Wredundant-decls so that gcc does not warn about this. This is re-enabled
  * at the bottom of this file.
  */
-#ifdef __GNUC__
+#ifdef __GNUC__ > 4
```
Comment (tom-cosgrove-arm) [inline library/constant_time_impl.h:34]: This isn't valid C pre-processor syntax, which should be `# ifdef identifier new-line`

**Comment by tom-cosgrove-arm** [comment]
Also, I'd note that gcc 4.5.2 was released on December 16, 2010, making it almost 13 years old

File: library/constant_time_impl.h (hunk 1 above)
Comment (mschulz-at-hilscher) [reply library/constant_time_impl.h:34]: sorry, that was a copy and paste mistake. I will fix that.

**Comment by mschulz-at-hilscher** [comment]
> Also, I'd note that gcc 4.5.2 was released on December 16, 2010, making it almost 13 years old

I agree, but it is not my choice to support that compiler. At least that little fix solves the problem.

File: library/constant_time_impl.h (hunk 1 above)
Comment (irwir) [reply library/constant_time_impl.h]: This fix may `pop` what was not pushed.

Comment (tom-cosgrove-arm) [reply library/constant_time_impl.h]: That would be best dealt with by something like we have in #8445

Comment (mschulz-at-hilscher) [reply library/constant_time_impl.h]: I now fixed it by fixing the check without the pop pragma.

File: library/constant_time_impl.h [hunk 2]
```diff
@@ -31,7 +31,7 @@
  * Disable -Wredundant-decls so that gcc does not warn about this. This is re-enabled
  * at the bottom of this file.
  */
-#ifdef __GNUC__
+#if __GNUC__ > 4
```
Comment (daverodgman) [inline library/constant_time_impl.h:34]: We should really check that the compiler is gcc - other compilers also define `__GNUC__`

```suggestion
#if defined(MBEDTLS_COMPILER_IS_GCC) && (__GNUC__ > 4)
```

File: library/constant_time_impl.h [hunk 3]
```diff
@@ -546,7 +546,7 @@ static inline mbedtls_ct_condition_t mbedtls_ct_bool_not(mbedtls_ct_condition_t
     return (mbedtls_ct_condition_t) (~x);
 }
 
-#ifdef __GNUC__
+#if __GNUC__ > 4
```
Comment (daverodgman) [inline library/constant_time_impl.h:549]:
```suggestion
#if defined(MBEDTLS_COMPILER_IS_GCC) && (__GNUC__ > 4)
```

**Review by daverodgman — CHANGES_REQUESTED** [review]

**Review by mschulz-at-hilscher — COMMENTED** [review]
Changes accepted

**Review by daverodgman — APPROVED** [review]

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Pull request merged**